cpu-o3: Clear thread state in time buffers on thread exit

nmosier opened the pull request:
Fix #1049. Clear only thread-specific state from the O3 CPU time buffers, rather than clearing *all* state for *all* threads.

Specifically, this patch adds a `clearStates()` method for all O3 time buffer structs in src/cpu/o3/comm.hh. Upon thread exit, `CPU::removeThread()` now invokes this method for all structs in each time buffer, rather than flushing out the time buffers (which nukes the states for all threads, not just the exiting one).

powerjg reviewed (commented):
This is a great solution. Thanks for fixing this issue!

I would prefer to see the functions in `comm.hh` to be in their respective places with the other member functions of each class. I do not believe we have any examples of this kind of style in gem5, and it will make future maintenance easier if we don't add new styles. Let me know if I'm wrong and we do this in other places, though.

Also, it would be good to have the change which modifies all of the initial values in cpu.hh to be in a different commit (no need to make a different PR, though).

erin-le commented:
@nmosier Hello, thank you for your contribution. Could you address Jason's comments when you have time? Thank you!

nmosier committed:
cpu-o3: add inline member initializers for DecodeComm, IewComm, CommitComm

Zero-initialize all members that didn't previously have default initializers
in TimeStruct::{DecodeComm,IewComm,CommitComm}.

Change-Id: I7004926ba94c4215be47efd61cc6f80400215c5c

nmosier committed:
cpu-o3: Clear thread state in time buffers on thread exit

Fix #1049. Clear only thread-specific state from the O3 CPU time buffers,
rather than clearing all state for all threads.

Change-Id: Ie8e6d91254c70081d931c7ab4632217b48bdbb78

nmosier commented:
@powerjg Thanks for the feedback! I tried to address it.

powerjg commented:
Apologies for the slow review!
